### ISSUE TITLE
Azure CLI task fails: AADSTS700024: Client assertion is not within its valid time range

### ISSUE THREAD

**Issue by jiasli** [Azure/azure-cli#28708]
Acquiring access token with expired OIDC token fails with:

> ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2024-04-05T23:01:54.2089203Z, assertion valid from 2024-04-05T22:40:41.0000000Z, expiry time of assertion 2024-04-05T22:50:41.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials

As the error indicates, the OIDC token is only valid for 10 minutes. After it is passed to `az login` via `--federated-token`, Azure CLI cannot get a new OIDC token after the OIDC token expires.

This is the designed v1 behavior of OIDC token support (#19853).

However, as Azure DevOps task **AzureCLI@2** (https://github.com/microsoft/azure-pipelines-tasks/pull/17633) and GitHub Action **azure/login@v2** (https://github.com/Azure/login/pull/147) have supported OIDC token authentication, and it is recommended to use [workload identity federation](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation), this limitation is becoming more prevailing.

## Possible solutions

1. OIDC token provider such as Azure DevOps or GitHub should provide an option to control the expiry time of the OIDC token to make it at least as long as the task duration.
2. Design and implement a v2 solution that uses a managed-identity-like interface which allows MSAL/Azure CLI to refresh OIDC token.

## References

- https://github.com/Azure/login/issues/180
- https://github.com/Azure/login/issues/372
- IcM 490937309
- IcM 491234676
- Email: Workload identity federation in Azure Pipelines fails with AADSTS700024

**Comment by yonzhan** [comment]
Thank you for opening this issue, we will look into it.

**Comment by jiasli** [comment]
## Callback interface proposals

Different external identity providers (IdP) have different ways of retrieving the ID token:

- GitHub Action exposes environment variable `ACTIONS_ID_TOKEN_REQUEST_URL` and `ACTIONS_ID_TOKEN_REQUEST_TOKEN` and requires a `GET` HTTP request: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-cloud-providers#requesting-the-jwt-using-environment-variables
- Azure DevOps exposes **Oidctoken - Create** API and requires a `POST` HTTP request: https://learn.microsoft.com/en-us/rest/api/azure/devops/distributedtask/oidctoken/create

I had a discussion with MSAL team today and proposed 2 possible callback interfaces:

1. Let each external IdP expose a callback command such as `getidtoken` that returns an ID token in `stdout`, then instead of providing `--federated-token <ID token>` to `az login`, they should provide `--federated-token-callback getidtoken` to `az login`, so that CLI and MSAL can actively retrieve an ID token with `getidtoken` when ID token expires.
2. Like the GitHub Action solution, define a manage-identity-like URL that can be used to get an ID token, such as `ID_TOKEN_REQUEST_URL`.

**Comment by jiasli** [comment]
## Mitigation: Extend task duration to 60 minutes

```
ID token:       |----| 10 min
Access token 1: |------------------------| 60 min
Access token 2:          | 20 min: ERROR: ID token expired
```

An ID token lasts for 5 minutes on GitHub Actions and 10 minutes on Azure DevOps, but an access token lasts for 60 minutes.

When you run `az login`, Azure CLI only acquires access tokens for ARM, using `https://management.core.windows.net//.scope` as the scope.

After the ID token expires, if acquiring an access token for other scopes, such as

```
az account get-access-token --scope https://kusto.kusto.windows.net//.default
```

as currently there is no access token for that scope in the token cache, Azure CLI/MSAL will try to get an access token with the ID token. However, as the ID token has expired, the command fails with `AADSTS700024`.

So, the mitigation is pretty straightforward: **Acquire all access tokens before the ID token expires.**

You have to know which scopes are used in your pipeline task and call `az account get-access-token --scope ...` immediately after `az login`. This makes Azure CLI/MSAL acquire access tokens for the specified scopes while the ID token is still valid and save them in the token cache. 

For example:

- Storage: `az account get-access-token --scope https://storage.azure.com/.default`
- Key Vault: `az account get-access-token --scope https://vault.azure.net/.default`
- Microsoft Graph: `az account get-access-token --scope https://graph.microsoft.com//.default`
- Kusto: `az account get-access-token --scope https://kusto.kusto.windows.net//.default`
 
Then subsequence commands using these scopes will use the access tokens saved in the token cache, so that they won't fail after the ID token expires, but they will still fail after the access token expires (60 minutes).

**Comment by mderriey** [comment]
Thanks for the mitigation @jiasli.

However, I don't think I'm hitting the issue where the Azure CLI tries to acquire an access token for a difference audience after the ID token has expired.

I'm fairly confident that the `az` commands I use only use the access token for ARM:

- `az account set`
- `az deployment sub create`
- `az deployment sub show`
- `az webapp deployment slot swap`
- `az webapp deployment source config-zip`
- `az webapp start`
- `az webapp stop`

The general flow is:

- Deploy an ARM template
- Deploy binaries to an App Service staging slot
- Swap slots
- Stop App Service staging slot

The time it takes to swap slots varies greatly, however more than 5 minutes have always elapsed by the time it's done.

Now, what is strange is that stopping the slot sometimes work, and sometimes doesn't, dependending on how much time has passed since we ran `azure/login`.

To me, it sounds like the access token expires "quicker" than before.
Could that be?

**Comment by jiasli** [comment]
@Kapsztajn, I can successfully get an access token for `https://servicebus.azure.net/.default` locally which lasts for 4600s.

```
> az account get-access-token --scope https://servicebus.azure.net/.default
{
  "accessToken": "...",
  "expiresOn": "2024-04-11 13:57:35.000000",
  "expires_on": 1712815055,
  "subscription": "0b1f6471-1bf0-4dda-aec3-cb9272f09590",
  "tenant": "54826b22-38d6-4fb2-bad9-b7b93a3e9c5a",
  "tokenType": "Bearer"
}
```

Decoded claims:

```
  "iat": 1712810455,
  "nbf": 1712810455,
  "exp": 1712815055,
```

I am not entirely sure why this line is printed:

```
CredentialUnavailableError: Please run 'az login' from a command prompt to authenticate before using this credential.
```

The [Azure Service Bus client library for JavaScript](https://www.npmjs.com/package/@azure/service-bus) SDK also didn't fail with `AADSTS700024`. I am not an expert of that SDK. Is it possible to collect more details on which scope the SDK requests, and why it fails with that error?

**Comment by jiasli** [comment]
@mderriey, this seems odd as all these operations are indeed ARM operations. Could you check the actual expiration time of the access token issued for ARM?

```
> az account get-access-token --scope https://management.core.windows.net//.default --query expiresOn --output tsv
2024-04-11 13:47:47.000000
```

**Comment by iamrk04** [comment]
Hi @Kapsztajn, the suggested mitigation did not work for me as well. It was able to fetch the token with an expiry that was reasonable, but I was able to see the same error once the OID token expired after 5 mins.

I propose a workaround by fetching the OID token every 4 mins to avoid the expiry. I was able to get this working and here is what I did: I inserted the following step in my workflow just before the step where this token expiry issue was popping:

```
      - name: Fetch OID token every 4 mins
        run: |
          while true; do
            token_request=$ACTIONS_ID_TOKEN_REQUEST_TOKEN
            token_uri=$ACTIONS_ID_TOKEN_REQUEST_URL
            token=$(curl -H "Authorization: bearer $token_request" "${token_uri}&audience=api://AzureADTokenExchange" | jq .value -r)
            az login --service-principal -u ${{ secrets.CLIENT_ID }} -t ${{ secrets.TENANT_ID }} --federated-token $token --output none
            # Sleep for 4 minutes
            sleep 240
          done &
```

Could you try this out and see if this works for you as well?

**Comment by mderriey** [comment]
> @mderriey, this seems odd as all these operations are indeed ARM operations. Could you check the actual expiration time of the access token issued for ARM?
> 
> ```
> > az account get-access-token --scope https://management.core.windows.net//.default --query expiresOn --output tsv
> 2024-04-11 13:47:47.000000
> ```

Good suggestion @jiasli , thanks.

Here's what I ran:

```yaml
steps:
- name: Login to Azure
  uses: azure/login@v2
  with:
    client-id: ${{ env.oidcAppRegistrationClientId }}
    tenant-id: ${{ env.azureTenantId }}
    allow-no-subscriptions: true
    enable-AzPSSession: true

- name: Check token expiry
  shell: bash
  run: |
    echo "Current date: $(date '+%Y-%m-%dT%H:%M:%S')"
    echo "Token expiration: $(az account get-access-token --resource-type arm --query expiresOn --output tsv --debug)"
    echo "Token #2 expiration: $(az account get-access-token --resource-type arm --query expiresOn --output tsv --debug)"
```

And the output (debug output omitted):

```console
Current date: 2024-04-11T06:57:14
Token expiration: 2024-04-11 07:57:14.000000
Token #2 expiration: 2024-04-11 07:57:14.000000
```

So the token is valid for 1 hour.

And both calls to `az account get-access-token` show this in the debug output, which I think confirms that the ARM token is cached and was originally acquired during `az login`:

```
DEBUG: msal.token_cache: event={
    "client_id": "***",
    "data": {
        "claims": "{\"access_token\": {\"xms_cc\": {\"values\": [\"CP1\"]}}}",
        "scope": [
            "https://management.core.windows.net//.default"
        ]
    },
    "environment": "login.microsoftonline.com",
    "grant_type": "client_credentials",
    "params": null,
    "response": {
        "access_token": "********",
        "expires_in": 3599,
        "ext_expires_in": 3599,
        "token_type": "Bearer"
    },
    "scope": [
        "https://management.core.windows.net//.default"
    ],
    "token_endpoint": "https://login.microsoftonline.com/<redacted>/oauth2/v2.0/token"
}
```

I'm not sure what happens, then...
I'll try removing the extra `azure/login` steps when I get some more time to see if the issue disappears.

Thanks again, let me know if I can perform some more testing if anything comes to mind.
If you'd be interested in the debug output, I could send that privately.

**Comment by ant0nsc** [comment]
@jiasli, thanks for suggesting this workaround. I tried your suggestion in my pipeline, but still run into the same issue as before. Example run: https://github.com/microsoft/hi-ml/actions/runs/8642139946/job/23692828663, using the workflow https://github.com/microsoft/hi-ml/pull/925/files#diff-efdacb46b092d7173f71346f14d9981f6697d11af5d8bfee360359810ba2eb17

Roughly speaking, in our test suite, we repeatedly run tests that
- get a credential (AzureCliCredential or service principal)
- run an Azure or AzureML operation
Because we run our test suite at the level of credential objects, it seems to me that we are not able to make use of your suggested workaround, which works at an access token level. (sorry if I'm getting things mixed up here, I'm certainly not an authentication expert).

**Comment by jiasli** [comment]
Apologize for the confusion caused.

As I tested today, the mitigation I provided in https://github.com/Azure/azure-cli/issues/28708#issuecomment-2047256166 is indeed not working for Azure CLI 2.59.0, because of an MSAL regression introduced in 1.27.0 (https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/688, https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/644) which is adopted by Azure CLI 2.59.0 (https://github.com/Azure/azure-cli/pull/28556). I will work with MSAL on this issue with high priority.

For now, please keep using **service principal secret** for authentication to get unblocked: https://github.com/marketplace/actions/azure-login#login-with-a-service-principal-secret

**Comment by Kapsztajn** [comment]
@iamrk04
It looks like your solution is working and I managed to run test normally (pipeline did run over 16 minutes). I have added code which you provide between Azure login and component test:

```
    - name: Azure login
      uses: azure/login@v2
      with:
        client-id: ${{ env.AZURE_CLIENT_ID }}
        tenant-id: ${{ env.AZURE_TENANT_ID }}
        subscription-id: ${{ env.AZURE_SUBSCRIPTION_ID }}
        enable-AzPSSession: false

    - name: Fetch OID token every 4 mins
      shell: bash
      run: |
        while true; do
          token_request=$ACTIONS_ID_TOKEN_REQUEST_TOKEN
          token_uri=$ACTIONS_ID_TOKEN_REQUEST_URL
          token=$(curl -H "Authorization: bearer $token_request" "${token_uri}&audience=api://AzureADTokenExchange" | jq .value -r)
          az login --service-principal -u ${{ env.AZURE_CLIENT_ID }} -t ${{ env.AZURE_TENANT_ID }} --federated-token $token --output none
          # Sleep for 4 minutes
          sleep 240
        done &

    - name: 'Run tests'
      shell: bash
      ...
```
I had to add `shell: bash` because without it I got errors with missing shell.

**Comment by smokedlinq** [comment]
@mderriey I assumed something like that, I was more referring to how that broke inside of `az`.

**Comment by ant0nsc** [comment]
> Hi @Kapsztajn, the suggested mitigation did not work for me as well. It was able to fetch the token with an expiry that was reasonable, but I was able to see the same error once the OID token expired after 5 mins.
> 
> I propose a workaround by fetching the OID token every 4 mins to avoid the expiry. I was able to get this working and here is what I did: I inserted the following step in my workflow just before the step where this token expiry issue was popping:
> 
> ```
>       - name: Fetch OID token every 4 mins
>         run: |
>           while true; do
>             token_request=$ACTIONS_ID_TOKEN_REQUEST_TOKEN
>             token_uri=$ACTIONS_ID_TOKEN_REQUEST_URL
>             token=$(curl -H "Authorization: bearer $token_request" "${token_uri}&audience=api://AzureADTokenExchange" | jq .value -r)
>             az login --service-principal -u ${{ secrets.CLIENT_ID }} -t ${{ secrets.TENANT_ID }} --federated-token $token --output none
>             # Sleep for 4 minutes
>             sleep 240
>           done &
> ```
> 
> Could you try this out and see if this works for you as well?

Hey @iamrk04, you're a hero! I inserted this snippet into my workflow, and this made it all work. Great idea to just have that run in the background in a shell loop.

For reference: https://github.com/microsoft/hi-ml/pull/925/

**Comment by avtakkar** [comment]
> Hi @Kapsztajn, the suggested mitigation did not work for me as well. It was able to fetch the token with an expiry that was reasonable, but I was able to see the same error once the OID token expired after 5 mins.
> 
> I propose a workaround by fetching the OID token every 4 mins to avoid the expiry. I was able to get this working and here is what I did: I inserted the following step in my workflow just before the step where this token expiry issue was popping:
> 
> ```
>       - name: Fetch OID token every 4 mins
>         run: |
>           while true; do
>             token_request=$ACTIONS_ID_TOKEN_REQUEST_TOKEN
>             token_uri=$ACTIONS_ID_TOKEN_REQUEST_URL
>             token=$(curl -H "Authorization: bearer $token_request" "${token_uri}&audience=api://AzureADTokenExchange" | jq .value -r)
>             az login --service-principal -u ${{ secrets.CLIENT_ID }} -t ${{ secrets.TENANT_ID }} --federated-token $token --output none
>             # Sleep for 4 minutes
>             sleep 240
>           done &
> ```
> 
> Could you try this out and see if this works for you as well?

Thanks @iamrk04 , this worked for me as well.

**Comment by nlighten** [comment]
Suggestion from @iamrk04 also worked for me. Wrapped it in a github action that potentially can replace `azure/login`. I think the solution will even remove the 1 hour limit we had before but have not tested this yet. 

```yaml
name: Azure Federated Login

inputs:
  client-id:
    description: Azure client id
    type: string
  tenant-id:
    description: Azure tenant id
    type: string
  subscription-id:
    description: Azure subscription id
    type: string
    default: none
  refresh-interval-seconds:
    description: Refresh interval in seconds
    type: number
    default: 240


runs:
  using: "composite"
  steps:
    - name: Fetch OID token every ${{ inputs.refresh-interval-seconds }} seconds
      shell: bash
      run: |
        first_time=true
        while true; do
          token=$(curl -s -H "Authorization: bearer ${ACTIONS_ID_TOKEN_REQUEST_TOKEN}" "${ACTIONS_ID_TOKEN_REQUEST_URL}&audience=api://AzureADTokenExchange" | jq .value -r)
          az login --service-principal -u ${{ inputs.client-id }} -t ${{ inputs.tenant-id }} --federated-token $token --output none
          if [ "$first_time" = true ] && [ "${{ inputs.subscription-id }}" != "none" ]; then
            az account set -s ${{ inputs.subscription-id }}
            first_time=false
          fi
          sleep ${{ inputs.refresh-interval-seconds }}
        done &
```

**Comment by TomWildenhain** [comment]
I'm running into the same issue in Azure Devops for a pipeline that runs a long python script (2h40m) in an AzureCLI@2 task. Was working fine on Friday (April 5th) but started failing after that with error:

`AzureCliCredential: ERROR: AADSTS700024: Client assertion is not within its valid time range. ... `

Any ideas on whether an equivalent workaround is possible for Azure Devops to refresh the token every 9 minutes?

**Comment by dghubble** [comment]
We started having problems with the v2.59.0 az cli and rolled back as a workaround. I'm not sure what about the cli release makes this more/less likely to hit this.

**Comment by jiasli** [comment]
> My question is why this has popped up as an issue recently. We've had pipelines run for well over 20 minutes before and never seen this. But within the last week, it seems any workflow using Azure CLI with OIDC federated auth is experiencing this issue.

@smokedlinq, please refer to my comment https://github.com/Azure/azure-cli/issues/28708#issuecomment-2049400226.

**Comment by jiasli** [comment]
> I propose a workaround by fetching the OID token every 4 mins to avoid the expiry.

This workaround https://github.com/Azure/azure-cli/issues/28708#issuecomment-2049014471 proposed by @iamrk04 of periodically calling `az login` is not recommended, as Azure CLI doesn't support concurrent execution and you will very likely run into some racing condition (https://github.com/Azure/azure-cli/issues/9427, https://github.com/Azure/azure-cli/issues/20273).

> We started having problems with the v2.59.0 az cli and rolled back as a workaround.

This workaround https://github.com/Azure/azure-cli/issues/28708#issuecomment-2050804548 proposed by @dghubble of using an old version is a correct one.

As I suggested in https://github.com/Azure/azure-cli/issues/28708#issuecomment-2049400226, using service principal secret for authentication is also another acceptable workaround.

**Comment by andre-qumulo** [comment]
@jiasli Service principals are unacceptable for some of us as our security certification would require we rotate them on a regular basis.  OIDC does not add that additional burden given that they are clearly short lived.

**Comment by jiasli** [comment]
I have created a separate issue https://github.com/Azure/azure-cli/issues/28737 to track "Azure CLI task fails with `AADSTS700024` after 5 minutes".

**Comment by TomWildenhain** [comment]
> I'm running into the same issue in Azure Devops for a pipeline that runs a long python script (2h40m) in an AzureCLI@2 task. Was working fine on Friday (April 5th) but started failing after that with error:
> 
> `AzureCliCredential: ERROR: AADSTS700024: Client assertion is not within its valid time range. ... `
> 
> Any ideas on whether an equivalent workaround is possible for Azure Devops to refresh the token every 9 minutes?

Thanks @jiasli! The mitigation steps for Azure DevOps provided [here](https://github.com/Azure/azure-cli/issues/28737#issuecomment-2051026300) of using a service principal secret were effective.

(I ran into some trouble finding the organization id while following the instructions but was able to find the organization id with these steps: https://medium.com/@shivapatel1102001/get-list-of-organization-from-azure-devops-microsoft-account-861ea29dae93)

**Comment by jiasli** [comment]
@TomWildenhain, based on my understanding, the steps provided by https://learn.microsoft.com/en-us/azure/devops/pipelines/library/connect-to-azure?view=azure-devops don't require organization ID when creating a service connection using service principal secret. Could you let me know which article you are following?

**Comment by geekzter** [comment]
@jiasli Org id is a 1P policy.

**Comment by TomWildenhain** [comment]
@jiasli Thanks for your help. I was following the instructions in a banner at the top of ADO after creating the manual service connection. The banner states:

 `Manually created service connections use an App Registration that was created by the user. Please add a federated credential to the App Registration with the following details: Issuer: https://vstoken.dev.azure.com/<org id>, Subject identifier: sc://<org>/<project>/<sc name>. Learn more`

With a link to: https://learn.microsoft.com/en-us/azure/devops/pipelines/release/configure-workload-identity?view=azure-devops

I used the instructions to call the API here to get the org id: https://medium.com/@shivapatel1102001/get-list-of-organization-from-azure-devops-microsoft-account-861ea29dae93

**Comment by jiasli** [comment]
@TomWildenhain, thanks for the information. If you used service principal secret to create the service connection, I don't think federated identity credential added to the app is actually used.

**Comment by nlighten** [comment]
@jiasli Is it possible to give any realistic timeline for a fix? I am wondering if it makes sense to ask for a roll back of the cli version contained in [actions/runner-images](https://github.com/actions/runner-images) that is used by both Github Actions and Azure DevOps.

**Comment by Moazzem-Hossain-pixel** [comment]
> Acquiring access token with expired OIDC token fails with:
> 
> > ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2024-04-05T23:01:54.2089203Z, assertion valid from 2024-04-05T22:40:41.0000000Z, expiry time of assertion 2024-04-05T22:50:41.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials
> 
> As the error indicates, the OIDC token is only valid for 10 minutes. After it is passed to `az login` via `--federated-token`, Azure CLI cannot get a new OIDC token after the OIDC token expires.
> 
> This is the designed v1 behavior of OIDC token support (#19853).
> 
> However, as Azure DevOps task **AzureCLI@2** ([microsoft/azure-pipelines-tasks#17633](https://github.com/microsoft/azure-pipelines-tasks/pull/17633)) and GitHub Action **azure/login@v2** ([Azure/login#147](https://github.com/Azure/login/pull/147)) have supported OIDC token authentication, and it is recommended to use [workload identity federation](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation), this limitation is becoming more prevailing.
> 
> ## Possible solutions
> 1. OIDC token provider such as Azure DevOps or GitHub should provide an option to control the expiry time of the OIDC token to make it at least as long as the task duration.
> 2. Design and implement a v2 solution that uses a managed-identity-like interface which allows MSAL/Azure CLI to refresh OIDC token.
> 
> ## References
> * [ERROR: AADSTS700024: Client assertion is not within its valid time range login#180](https://github.com/Azure/login/issues/180)
> * [ERROR: AADSTS700024: Client assertion is not within its valid time range login#372](https://github.com/Azure/login/issues/372)
> * IcM 490937309
> * IcM 491234676
> * Email: Workload identity federation in Azure Pipelines fails with AADSTS700024

**Comment by Moazzem-Hossain-pixel** [comment]
If you can take the required initiative, will be surprised.

**Comment by Moazzem-Hossain-pixel** [comment]
> ## Mitigation: Extend task duration to 60 minutes
> Warning
> 
> This mitigation doesn't work with Azure CLI 2.59.0. See [#28708 (comment)](https://github.com/Azure/azure-cli/issues/28708#issuecomment-2049400226).
> 
> ```
> ID token:       |----| 10 min
> Access token 1: |------------------------| 60 min
> Access token 2:          | 20 min: ERROR: ID token expired
> ```
> 
> An ID token lasts for 5 minutes on GitHub Actions and 10 minutes on Azure DevOps, but an access token lasts for 60 minutes.
> 
> When you run `az login`, Azure CLI only acquires access tokens for ARM, using `https://management.core.windows.net//.default` as the scope.
> 
> After the ID token expires, if acquiring an access token for other scopes, such as
> 
> ```
> az account get-access-token --scope https://kusto.kusto.windows.net//.default
> ```
> 
> as currently there is no access token for that scope in the token cache, Azure CLI/MSAL will try to get an access token with the ID token. However, as the ID token has expired, the command fails with `AADSTS700024`.
> 
> So, the mitigation is pretty straightforward: **Acquire all access tokens before the ID token expires.**
> 
> You have to know which scopes are used in your pipeline task and call `az account get-access-token --scope ...` immediately after `az login`. This makes Azure CLI/MSAL acquire access tokens for the specified scopes while the ID token is still valid and save them in the token cache.
> 
> For example:
> 
> * Storage: `az account get-access-token --scope https://storage.azure.com/.default --output none`
> * Key Vault: `az account get-access-token --scope https://vault.azure.net/.default --output none`
> * Microsoft Graph: `az account get-access-token --scope https://graph.microsoft.com//.default --output none`
> * Kusto: `az account get-access-token --scope https://kusto.kusto.windows.net//.default --output none`
> 
> Warning
> 
> Even though GitHub Actions can mask the access token as `***` in `az account get-access-token`'s output:
> 
> ```
> + az account get-access-token
> ***
>   "accessToken": "***",
>   "expiresOn": "2024-04-10 14:11:25.000000",
>   "expires_on": 1712758285,
>   "subscription": "...",
>   "tenant": "...",
>   "tokenType": "Bearer"
> ***
> ```
> 
> **You MUST specify `--output none` to make sure no access token is printed to any of your logs.**
> 
> Then subsequence commands using these scopes will use the access tokens saved in the token cache, so that they won't fail after the ID token expires, but they will still fail after the access token expires (60 minutes).

**Comment by Moazzem-Hossain-pixel** [comment]
Need help

**Comment by pkoushik** [comment]
We are seeing the same issue related to moving away from service principal secrets. 

We are looking into adding logic for all Az CLI calls using the ARM token to ensure it gets refreshed (but not as a background process) to get the OIDC token from `idToken` and reuse it to log in via `az account clear && az login ...`

**Comment by panpanwa** [comment]
I have exactly the same use case as @TomWildenhain. Is there a way to make the token valid period customable?  We can't use Service principal as that's discouraged by the cred free best practices.

Even a workaround would be much appreciated.

**Comment by jhwj9617** [comment]
Have the same issue for our long-running tasks:
```
[01:50:31 INF]  ---> (Inner Exception #3) Azure.Identity.CredentialUnavailableException: Azure CLI authentication failed due to an unknown error. See the troubleshooting guide for more information. https://aka.ms/azsdk/net/identity/azclicredential/troubleshoot ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2024-06-01T01:50:31.1765304Z, assertion valid from 2024-06-01T00:49:55.0000000Z, expiry time of assertion 2024-06-01T00:59:55.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials . Trace ID: 48af9e38-7793-458d-94af-c2962d617700 Correlation ID: 0f495332-706e-4dba-a18e-1f844f5d7a7d Timestamp: 2024-06-01 01:50:31Z
[01:50:31 INF] Interactive authentication is needed. Please run:
[01:50:31 INF] az login
```

**Comment by panpanwa** [comment]
@jhwj9617 refer to the solution provided by [Kapsztajn](https://github.com/Kapsztajn) and @iamrk04, it works for me too.

Although I believe this is an unnecessary workaround which has to be done by users!

**Comment by jhwj9617** [comment]
@panpanwa we are not using github actions. We're using AzureDevOps in yml, e.g.
```yml
- task: AzureCLI@2
  displayName: Run load profile
  inputs:
    azureSubscription: $(federatedCredConnection)
    scriptType: ps
    scriptLocation: scriptPath
    scriptPath: $(Pipeline.Workspace)/test.ps1
```

**Comment by jhwj9617** [comment]
Also this seems to be in preview for [v1.12.0-beta.2](https://www.nuget.org/packages/Azure.Identity/1.12.0-beta.2)
https://github.com/Azure/azure-sdk-for-js/pull/29392

**Comment by kboom** [comment]
> @panpanwa this is the stopgap solution that was shared by a colleague we can implement in our AzureCLI task
> 
> ```powershell
> Start-Job -Name 'RefreshOidcToken' -ScriptBlock {
>   do {
>     Get-ChildItem -Path Env: -Recurse -Include ENDPOINT_DATA_* `
>       | Select-Object -First 1 -ExpandProperty Name `
>       | ForEach-Object { $_.Split("_")[2] } `
>       | Set-Variable serviceConnectionId
> 
>     $oidcRequestUrl = "${env:SYSTEM_TEAMFOUNDATIONCOLLECTIONURI}${env:SYSTEM_TEAMPROJECTID}/_apis/distributedtask/hubs/build/plans/${env:SYSTEM_PLANID}/jobs/${env:SYSTEM_JOBID}/oidctoken?api-version=7.1-preview.1&serviceConnectionId=${serviceConnectionId}"
>     Invoke-RestMethod -Headers @{
>       Authorization  = "Bearer $env:SYSTEM_ACCESSTOKEN"
>       'Content-Type' = 'application/json'
>     } -Uri "${oidcRequestUrl}" -Method Post | Set-Variable oidcTokenResponse
> 
>     $oidcToken = $oidcTokenResponse.oidcToken
>     if (!$oidcToken) {
>       Write-Warning "OIDC token could not be acquired. Retrying..."
>       Start-Sleep -Seconds 30
>       continue
>     }
> 
>     az account show -o json | ConvertFrom-Json | Set-Variable account
>     az login --service-principal -u $account.user.name --tenant $account.tenantId --allow-no-subscriptions --federated-token $oidcToken | Out-Null
> 
>     Start-Sleep -Seconds 480 # 8 minutes
>   } while ($true)
> } | Tee-Object -Variable refreshOidcTokenJob `
>   | Select-Object -ExcludeProperty Command `
>   | Write-Host -ForegroundColor DarkMagenta
> 
> # do long running work
> 
> Receive-Job $refreshOidcTokenJob
> Stop-Job -Job $refreshOidcTokenJob
> Remove-Job -Job $refreshOidcTokenJob
> ```

This might work in 99% of the cases but is not completely reliable; beware of race conditions.

**Comment by jiasli** [comment]
Azure DevOps's document now also explains `AADSTS700024`:

https://learn.microsoft.com/en-us/azure/devops/pipelines/release/troubleshoot-workload-identity

> AADSTS700024: Client assertion is not within its valid time range
>
> If the error happens after approximately 1 hour, use a service connection with [Workload identity federation and a Managed Identity](https://learn.microsoft.com/en-us/azure/devops/pipelines/release/configure-workload-identity?view=azure-devops#set-a-workload-identity-service-connection-to-use-managed-identity-authentication) instead. Managed Identity tokens have a [lifetime of around 24 hours](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/managed-identities-faq#are-managed-identities-tokens-cached).
If the error happens before 1 hour but after 10 minutes, move commands that (implicitly) request an access token to e.g. access Azure storage to the beginning of your script. The access token will be cached for subsequent commands.

**Comment by jluongh** [comment]
Do we have any updates on the issue? A lot of our ADO pipelines are intermittently failing and we have been asked to move away from service principals to be cred free.

The PR linked is still in draft state https://github.com/Azure/azure-cli/pull/28778

**Comment by TomWildenhain** [comment]
> Azure DevOps's document now also explains `AADSTS700024`:
> 
> https://learn.microsoft.com/en-us/azure/devops/pipelines/release/troubleshoot-workload-identity
> 
> > AADSTS700024: Client assertion is not within its valid time range
> > If the error happens after approximately 1 hour, use a service connection with [Workload identity federation and a Managed Identity](https://learn.microsoft.com/en-us/azure/devops/pipelines/release/configure-workload-identity?view=azure-devops#set-a-workload-identity-service-connection-to-use-managed-identity-authentication) instead. Managed Identity tokens have a [lifetime of around 24 hours](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/managed-identities-faq#are-managed-identities-tokens-cached).
> > If the error happens before 1 hour but after 10 minutes, move commands that (implicitly) request an access token to e.g. access Azure storage to the beginning of your script. The access token will be cached for subsequent commands.

Thanks @jiasli! This works for my use case!

**Comment by nrv-96** [comment]
I got same error for the time duration between 10 min to 1 hour, as mentioned on the [Microsoft Docs](https://learn.microsoft.com/en-us/azure/devops/pipelines/release/troubleshoot-workload-identity?view=azure-devops#:~:text=If%20the%20error%20happens%20before%201%20hour%20but%20after%2010%20minutes%2C%20move%20commands%20that%20(implicitly)%20request%20an%20access%20token%20to%20e.g.%20access%20Azure%20storage%20to%20the%20beginning%20of%20your%20script.%20The%20access%20token%20will%20be%20cached%20for%20subsequent%20commands.)  as mentioned in the docs we have access storage account at beginning but in terraform apply we cannot manage by ourselves.

I'm using terraform apply the pipeline running around 10 min and then gives below error: 

> error loading state: Error retrieving keys for Storage Account "teestmgmt": autorest/Client#Do: Preparing request failed: StatusCode=0 -- Original Error: clientCredentialsToken: received HTTP status 401 with response: {"error":"invalid_client","error_description":"AADSTS700024: Client assertion is not within its valid time range. Current time: 2024-06-21T10:41:11.0510669Z, assertion valid from 2024-06-19T02:32:13.0000000Z, expiry time of assertion 2024-06-19T02:42:13.0000000Z. Review the documentation at 
> https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials . Trace ID:  Correlation ID:  Timestamp: 2024-06-21 10:41:11Z","error_codes":[700024],"timestamp":"2024-06-21 10:41:11Z","trace_id":"","correlation_id":"","error_uri":"https://login.microsoftonline.com/error?code=700024"}

**Comment by chriswue** [comment]
Just to be clear - this isn't just a problem with github runners, happens on AzDO pipelines as well. Sprinkling `az account get-access-token -o none` all over our deployment scripts works but is rather ugly.

**Comment by aloizeau** [comment]
> I got same error for the time duration between 10 min to 1 hour, as mentioned on the [Microsoft Docs](https://learn.microsoft.com/en-us/azure/devops/pipelines/release/troubleshoot-workload-identity?view=azure-devops#:~:text=If%20the%20error%20happens%20before%201%20hour%20but%20after%2010%20minutes%2C%20move%20commands%20that%20(implicitly)%20request%20an%20access%20token%20to%20e.g.%20access%20Azure%20storage%20to%20the%20beginning%20of%20your%20script.%20The%20access%20token%20will%20be%20cached%20for%20subsequent%20commands.) as mentioned in the docs we have access storage account at beginning but in terraform apply we cannot manage by ourselves.
> 
> I'm using terraform apply the pipeline running around 10 min and then gives below error:
> 
> > error loading state: Error retrieving keys for Storage Account "teestmgmt": autorest/Client#Do: Preparing request failed: StatusCode=0 -- Original Error: clientCredentialsToken: received HTTP status 401 with response: {"error":"invalid_client","error_description":"AADSTS700024: Client assertion is not within its valid time range. Current time: 2024-06-21T10:41:11.0510669Z, assertion valid from 2024-06-19T02:32:13.0000000Z, expiry time of assertion 2024-06-19T02:42:13.0000000Z. Review the documentation at
> > https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials . Trace ID:  Correlation ID:  Timestamp: 2024-06-21 10:41:11Z","error_codes":[700024],"timestamp":"2024-06-21 10:41:11Z","trace_id":"","correlation_id":"","error_uri":"[https://login.microsoftonline.com/error?code=700024"}](https://login.microsoftonline.com/error?code=700024%22%7D)

I've got the same issue using terraform deployment and federated identity by an user assigned identity. 
- Do you have any idea to renew the token in this use case ? 
I see this article: https://learn.microsoft.com/en-us/entra/identity-platform/configurable-token-lifetimes but 

> _We currently don't support configuring the token lifetimes for [managed identity service principals](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/overview)._

**Comment by chriswue** [comment]
> Just to be clear - this isn't just a problem with github runners, happens on AzDO pipelines as well. Sprinkling `az account get-access-token -o none` all over our deployment scripts works but is rather ugly.

I stand corrected - we still see intermittent failures. Basically, federated identity is unusable at this stage as long as you have slightly more complex build scenarios that require more than 10min to deploy. We have converted back to service principal + secret

**Comment by Fl-Ta** [comment]
Perhaps interesting as a workaround:

keepAzSessionActive: true

[https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/AzureCLIV2/azureclitask.ts](url)


```
  var keepAzSessionActive: boolean = tl.getBoolInput('keepAzSessionActive', false);
            var stopRefreshingSession: () => void = () => {};
            if (keepAzSessionActive) {
                // This is a tactical workaround to keep the session active for the duration of the task to avoid AADSTS700024 errors.
                // This is a temporary solution until the az cli provides a way to refresh the session.
                if (authorizationScheme !== 'workloadidentityfederation') {
                    const errorMessage = tl.loc('KeepingAzSessionActiveUnsupportedScheme', authorizationScheme);
                    tl.error(errorMessage);
                    throw errorMessage;
                }
                stopRefreshingSession = this.keepRefreshingAzSession(connectedService);
            }
```

**Comment by vanhoutenbos** [comment]
We have the same situation as the OP.

We run a long-running task (making a backup) that sometimes runs longer than 1 hour.
We poll the status of the backup and wait for a 'success' before proceeding.

using `az account get-access-token --resource=https://management.azure.com/ --query accessToken -o tsv` to get a new refresh token works as long as the access token hasnt expired (60 minutes).
We call this function every 15 minutes, which works 3 times, but the 4th time is after the 1 hour mark and returns the same error as the OP mentions.

We run this inside a `AzureCLI@2` task in Azure DevOps.
How can we extend the lifetime or create a new access token without leaving the pipeline?

Do we need to call `az login` again?

**Comment by zioalex** [comment]
This is happening also with Terraform from ADO before 60mins. It fails consistently after ~45mins.

provider.terraform-provider-azurerm_v3.90.0_x5
and Terraform 1.10.4

The operation was an image promotion step:

```
azurerm_shared_image_version.image_new_version: Still creating... [43m30s elapsed]
2025-01-14T13:07:27.755Z [ERROR] provider.terraform-provider-azurerm_v3.90.0_x5: Response contains error diagnostic: @caller=github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/diag/diagnostics.go:58 @module=sdk.proto tf_proto_version=5.4 tf_req_id=a36a40c5-f7d6-4e8c-b764-c94985033488 tf_rpc=ApplyResourceChange diagnostic_detail="" diagnostic_severity=ERROR diagnostic_summary="waiting for the creation of Shared Image Version: (Version Name \"999.999.999\" / Image Name \"dev-img-win-2022\" / Gallery Name \"dev_imgg\" / Resource Group \"dev-rg\"): autorest/Client#Do: Preparing request failed: StatusCode=0 -- Original Error: clientCredentialsToken: received HTTP status 401 with response: {\"error\":\"invalid_client\",\"error_description\":\"AADSTS700024: Client assertion is not within its valid time range. Current time: 2025-01-14T13:07:27.7294524Z, assertion valid from 2025-01-14T12:23:32.0000000Z, expiry time of assertion 2025-01-14T12:33:32.0000000Z. Review the documentation at https://learn.microsoft.com/entra/identity-platform/certificate-credentials . "error_codes\":[700024],\"timestamp\":\"2025-01-14 13:07:27Z\",\"error_uri\":\"[https://login.microsoftonline.com/error?code=700024\"}"](https://login.microsoftonline.com/error?code=700024\%22}%22) tf_provider_addr=provider tf_resource_type=azurerm_shared_image_version timestamp=2025-01-14T13:07:27.755Z
2025-01-14T13:07:27.755Z [ERROR] vertex "azurerm_shared_image_version.image_new_version" error: waiting for the creation of Shared Image Version: (Version Name "999.999.999" / Image Name "dev-img-win-2022" / Gallery Name "dev_imgg" / Resource Group "dev-rg"): autorest/Client#Do: Preparing request failed: StatusCode=0 -- Original Error: clientCredentialsToken: received HTTP status 401 with response: {"error":"invalid_client","error_description":"AADSTS700024: Client assertion is not within its valid time range. Current time: 2025-01-14T13:07:27.7294524Z, assertion valid from 2025-01-14T12:23:32.0000000Z, expiry time of assertion 2025-01-14T12:33:32.0000000Z. Review the documentation at https://learn.microsoft.com/entra/identity-platform/certificate-credentials . Timestamp: 2025-01-14 13:07:27Z","error_codes":[700024],"timestamp":"2025-01-14 13:07:27Z","error_uri":"[https://login.microsoftonline.com/error?code=700024"}](https://login.microsoftonline.com/error?code=700024%22})
```

**Comment by Poltergeisen** [comment]
It's still occurring for terraform tasks, very frustrating for end-users and long running pipelines

**Comment by 4c74356b41** [comment]
its been about a year, any plans on actually fixing this? this is unacceptable

**Comment by alokjain-01** [comment]
> Acquiring access token with expired OIDC token fails with:
> 
> > ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2024-04-05T23:01:54.2089203Z, assertion valid from 2024-04-05T22:40:41.0000000Z, expiry time of assertion 2024-04-05T22:50:41.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials
> 
> As the error indicates, the OIDC token is only valid for 10 minutes. After it is passed to `az login` via `--federated-token`, Azure CLI cannot get a new OIDC token after the OIDC token expires.
> 
> This is the designed v1 behavior of OIDC token support ([#19853](https://github.com/Azure/azure-cli/pull/19853)).
> 
> However, as Azure DevOps task **AzureCLI@2** ([microsoft/azure-pipelines-tasks#17633](https://github.com/microsoft/azure-pipelines-tasks/pull/17633)) and GitHub Action **azure/login@v2** ([Azure/login#147](https://github.com/Azure/login/pull/147)) have supported OIDC token authentication, and it is recommended to use [workload identity federation](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation), this limitation is becoming more prevailing.
> 
> ## Possible solutions
> 1. OIDC token provider such as Azure DevOps or GitHub should provide an option to control the expiry time of the OIDC token to make it at least as long as the task duration.
> 2. Design and implement a v2 solution that uses a managed-identity-like interface which allows MSAL/Azure CLI to refresh OIDC token.
> 
> ## References
> * [ERROR: AADSTS700024: Client assertion is not within its valid time range login#180](https://github.com/Azure/login/issues/180)
> * [ERROR: AADSTS700024: Client assertion is not within its valid time range login#372](https://github.com/Azure/login/issues/372)
> * IcM 490937309
> * IcM 491234676
> * Email: Workload identity federation in Azure Pipelines fails with AADSTS700024



> Acquiring access token with expired OIDC token fails with:
> 
> > ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2024-04-05T23:01:54.2089203Z, assertion valid from 2024-04-05T22:40:41.0000000Z, expiry time of assertion 2024-04-05T22:50:41.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials
> 
> As the error indicates, the OIDC token is only valid for 10 minutes. After it is passed to `az login` via `--federated-token`, Azure CLI cannot get a new OIDC token after the OIDC token expires.
> 
> This is the designed v1 behavior of OIDC token support ([#19853](https://github.com/Azure/azure-cli/pull/19853)).
> 
> However, as Azure DevOps task **AzureCLI@2** ([microsoft/azure-pipelines-tasks#17633](https://github.com/microsoft/azure-pipelines-tasks/pull/17633)) and GitHub Action **azure/login@v2** ([Azure/login#147](https://github.com/Azure/login/pull/147)) have supported OIDC token authentication, and it is recommended to use [workload identity federation](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation), this limitation is becoming more prevailing.
> 
> ## Possible solutions
> 1. OIDC token provider such as Azure DevOps or GitHub should provide an option to control the expiry time of the OIDC token to make it at least as long as the task duration.
> 2. Design and implement a v2 solution that uses a managed-identity-like interface which allows MSAL/Azure CLI to refresh OIDC token.
> 
> ## References
> * [ERROR: AADSTS700024: Client assertion is not within its valid time range login#180](https://github.com/Azure/login/issues/180)
> * [ERROR: AADSTS700024: Client assertion is not within its valid time range login#372](https://github.com/Azure/login/issues/372)
> * IcM 490937309
> * IcM 491234676
> * Email: Workload identity federation in Azure Pipelines fails with AADSTS700024

This is very much needed since as per the latest security advice , OIDC authentication token based approach is the recommended approach. We are using the OIDC authentication in our ADO pipeline and in the pipeline jobs, we are using terraform templates which is used for the resource deployment. We are using OIDC token for the backend azure authentication. The resource deployment job can runs into hours ( for large number of resources deployment ) , But in our case the job is terminating abruptly as the OIDC token validity is for 10 minutes only We tried the suggestions given in the thread  to increase it to validity upto 60 minute but it does not work. What we need the OIDC token expiration time to be user define for the job so that we can set it to as long as it requires for the the job to run successfully.

**Comment by Poltergeisen** [comment]
@jamesadevine have you tried the auth method with environment variables? I think we switched to start using that flag and we get a lot fewer token errors. The refresh token gets saved in the plan itself and that's why it doesn't always work I believe

**Comment by 4c74356b41** [comment]
which method? @jamesadevine there's no mentions of such method.

**Comment by jamesadevine** [comment]
> [@jamesadevine](https://github.com/jamesadevine) have you tried the auth method with environment variables? I think we switched to start using that flag and we get a lot fewer token errors. The refresh token gets saved in the plan itself and that's why it doesn't always work I believe

Is this where you add use addSpnToEnvironment, save the env vars as pipeline variables, and then call az login in a separate step outside of the context of the AzureCLI@2 task?

**Comment by stevehipwell** [comment]
How come the CLI doesn't support using the `ACTIONS_ID_TOKEN_REQUEST_TOKEN` & `ACTIONS_ID_TOKEN_REQUEST_URL` env variables provided by GitHub? Both the [AzureRM TF provider](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs) and [Kubelogin](https://github.com/Azure/kubelogin) already do this.

As a workaround we use the `--federated-token` but keep this refreshed with a background task.

**Comment by geekzter** [comment]
@Poltergeisen The new TerraformTask@5 supports token refresh. See [Introducing Azure DevOps ID Token Refresh and Terraform Task Version 5](https://devblogs.microsoft.com/devops/introducing-azure-devops-id-token-refresh-and-terraform-task-version-5/).

**Comment by TimEllens** [comment]
Also having this issue when running Terraform Test.
Implemented the suggestions from https://devblogs.microsoft.com/devops/introducing-azure-devops-id-token-refresh-and-terraform-task-version-5/ for AzureCLI@2, but still get the error. 

`- task: AzureCLI@2
    displayName: 'Terraform'
    inputs:
      azureSubscription: my-service-connection
      scriptType: pscore
      scriptLocation: inlineScript
      inlineScript: |
        $subscriptionId = $(az account show --query id -o tsv)

        $env:ARM_TENANT_ID = $env:AZURESUBSCRIPTION_TENANT_ID
        $env:ARM_SUBSCRIPTION_ID = $subscriptionId
        $env:ARM_CLIENT_ID =  $env:AZURESUBSCRIPTION_CLIENT_ID
        $env:ARM_OIDC_AZURE_SERVICE_CONNECTION_ID = $env:AZURESUBSCRIPTION_SERVICE_CONNECTION_ID
        $env:ARM_USE_OIDC = "true"

        terraform init `
          -backend-config="storage_account_name=my-storage-accont" `
          -backend-config="container_name=my-container" `
          -backend-config="key=terraform.tfstate" `
          -backend-config="use_azuread_auth=true"

        terraform plan -out tfplan
        terraform apply tfplan
    env:
      ARM_OIDC_REQUEST_TOKEN: $(System.AccessToken)`

**Comment by lbpphr** [comment]
> How come the CLI doesn't support using the `ACTIONS_ID_TOKEN_REQUEST_TOKEN` & `ACTIONS_ID_TOKEN_REQUEST_URL` env variables provided by GitHub? Both the [AzureRM TF provider](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs) and [Kubelogin](https://github.com/Azure/kubelogin) already do this.
> 
> As a workaround we use the `--federated-token` but keep this refreshed with a background task.

@stevehipwell how are you doing that exactly?

I'm using an `az sql db import` command with a managed identity and running into the same issue.

**Comment by stevehipwell** [comment]
@lbpphr in essence I create a bash script to run `curl -H "Authorization: bearer ${ACTIONS_ID_TOKEN_REQUEST_TOKEN}" "${ACTIONS_ID_TOKEN_REQUEST_URL}&audience=api://AzureADTokenExchange"` in the background every 4 mins ([OIDC docs](https://docs.github.com/en/actions/concepts/security/about-security-hardening-with-openid-connect#updating-your-actions-for-oidc)).

**Comment by jenshnielsen** [comment]
For me using `keepAzSessionActive` as suggested by @Fl-Ta resolved the issue. It would be nice it this argument to AzureCLI@2 was documented here https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/azure-cli-v2?view=azure-pipelines. That would have saved me some time

**Comment by hollowdrutt** [comment]
> You have to know which scopes are used in your pipeline task and call `az account get-access-token --scope ...` immediately after `az login`. This makes Azure CLI/MSAL acquire access tokens for the specified scopes while the ID token is still valid and save them in the token cache.
> 
> For example:
> 
> * Storage: `az account get-access-token --scope https://storage.azure.com/.default --output none`
> * Key Vault: `az account get-access-token --scope https://vault.azure.net/.default --output none`
> * Microsoft Graph: `az account get-access-token --scope https://graph.microsoft.com//.default --output none`
> * Kusto: `az account get-access-token --scope https://kusto.kusto.windows.net//.default --output none`

@jiasli How do I know what scope to use for what Azure service? I had issues with trusted signing and guessing from your examples I used `az account get-access-token --scope https://codesigning.azure.net/.default --output none` which seems to work, but the only reference to that scope I could find was in a blog post by Scott Hanselman: 
https://www.hanselman.com/blog/automatically-signing-a-windows-exe-with-azure-trusted-signing-dotnet-sign-and-github-actions

Are these scopes documented somewhere?
